src/azure-cli/azure/cli/command_modules/sqlvm/_validators.py

```diff
@@ -186,3 +190,322 @@ def validate_assessment_start_time_local(namespace):
+
+    for umi in vm.identity.user_assigned_identities.values():
+        if umi.client_id == namespace.msi_client_id:
+            return umi.principal_id
+
+    az_error = InvalidArgumentValueError("Enable Azure AD authentication with user-assigned managed identity {}, but the managed identity is not attached to this Azure virtual machine.".format(namespace.msi_client_id))
+    az_error.set_recommendation("Attach the user-assigned managed identity {} to the Azure virtual machine {}.".format(namespace.msi_client_id, namespace.sql_virtual_machine_name))
+    raise az_error
+
+
+# Validate the MSI has appropriate permissions to query Microsoft Graph API.
+USER_READ_ALL = "User.Read.All"
+APPLICATION_READ_ALL = "Application.Read.All"
+GROUP_MEMBER_READ_ALL = "GroupMember.Read.All"
+
+
+def _validate_msi_with_enough_permission(cli_ctx, principal_id):
+    """ Validate the MSI has enough permissions to query Microsoft Graph API, which are needed for SQL server to
+        carry out the Azure AD authentication.
+
+        :param cli_ctx: The CLI context.
+        :type cli_ctx: AzCli.
+        :param principal_id: The principalId of the MSI.
+        :type principal_id: str.
+    """
+    logger.debug("Validate Azure AD authentication regarding required Graph API permission.")
+
+    directory_roles = _directory_role_list(cli_ctx, principal_id)
+
+    # If the MSI is assigned the "Directory Readers" role, it has enough permissions.
+    if any(role["displayName"] == "Directory Readers" for role in directory_roles):
+        return
+
+    # If the MSI is not assigned the "Directory Readers" role, check the app roles.
+    # Retrieve the app role Id for User.Read.All, Application.Read.All, GroupMember.Read.All roles.
+    app_role_id_map = _find_role_id(cli_ctx)
+    logger.debug("Validate Azure AD authentication: app role to app role Id map:%s.", str(app_role_id_map))
+
+    # Retrieve all the role assignments assigned to the MSI
+    app_role_assignments = _app_role_assignment_list(cli_ctx, principal_id)
+    all_assigned_role_ids = [assignment["appRoleId"] for assignment in app_role_assignments]
+
+    # Find all the missing roles.
+    required_role_names = [USER_READ_ALL, APPLICATION_READ_ALL, GROUP_MEMBER_READ_ALL]
+    missing_roles = [role_name for role_name in required_role_names if app_role_id_map[role_name] not in all_assigned_role_ids]
+
+    if len(missing_roles) > 0:
+        az_error = InvalidArgumentValueError("The managed identity is lack of the following roles for Azure AD authentication: {}.".format(", ".join(missing_roles)))
+        az_error.set_recommendation("Grant the managed identity EITHER the Directory.Readers role OR the three App roles 'User.Read.All', 'Application.Read.All', 'GroupMember.Read.All'")
+        raise az_error
+
+
+MICROSOFT_GRAPH_API_ERROR = "Unable to validate the permission of MSI due to querying Microsoft Graph API encountered error: {}."
+
+
+def _send(cli_ctx, method, url, param=None, body=None):
```

Comment:
I see a lot of duplicated code here with [Graph Client in role module](https://github.com/Azure/azure-cli/blob/dev/src/azure-cli/azure/cli/command_modules/role/_msgrpah/_graph_client.py)

cc @jiasli to confirm if it's best practice to keep code separately in two modules

Reply:
there was a discussion about this, I have added you to the email thread.